server/s3_put_multiobject_copy_action.cc

```diff
@@ -146,16 +146,123 @@ void S3PutMultipartCopyAction::validate_multipart_partcopy_request() {
     send_response_to_s3_client();
     return;
   }
-  if (MaxPartCopySourcePartSize <
-      additional_object_metadata->get_content_length()) {
-    s3_copy_part_action_state = S3PutObjectActionState::validationFailed;
-    set_s3_error("InvalidRequest");
-    send_response_to_s3_client();
+  std::string range_header_value =
+      request->get_header_value("x-amz-copy-source-range");
+  source_object_size = additional_object_metadata->get_content_length();
+
+  if (range_header_value.empty()) {
+    // Range is not specified, read complete object
+    s3_log(S3_LOG_DEBUG, request_id, "Range is not specified\n");
+    if (MaxPartCopySourcePartSize < source_object_size) {
+      s3_copy_part_action_state = S3PutObjectActionState::validationFailed;
+      set_s3_error("InvalidRequest");
+      send_response_to_s3_client();
+    } else {
+      last_byte_offset_to_copy = source_object_size - 1;
+    }
   } else {
-    total_data_to_copy = additional_object_metadata->get_content_length();
-    next();
+    // parse the Range header value
+    // eg: bytes=0-1024 value
+    s3_log(S3_LOG_DEBUG, request_id, "Range found(%s)\n",
+           range_header_value.c_str());
+    if (!validate_range_header(range_header_value)) {
+      set_s3_error("InvalidRange");
+      send_response_to_s3_client();
+    }
+  }
+  total_data_to_copy = last_byte_offset_to_copy - first_byte_offset_to_copy;
```

Comment:
Why 'first_byte_offset_to_copy' & 'last_byte_offset_to_copy ' are kept global, in header file?Not a good way of sharing these variables, rather pass as ptrs it in function itself.

Reply:
These are not global variables. These are class variables, member of `S3PutMultipartCopyAction` class and they are private by default.

Reply:
Sorry, i just saw them in header file and didnt expand previous code to see its location.